### PR TITLE
JITSU-164: retry Data API throttles wrapped in non-retryable errors

### PR DESCRIPTION
**Pull request by absorbb**
Second round on the Redshift CI throttling (`JITSU-164`), prompted by [this run](https://github.com/jitsucom/jitsu/actions/runs/31819911512): failures persisted **after** #1462 because the actual error is

```
StatusCode: 400 ... ValidationException: Service returned error code ThrottlingException (Service: RedshiftServerless, Status Code: 400, ...)
```

The Data API wraps the underlying serverless service's throttle inside a 400 `ValidationException` — a *client error* class the AWS SDK retryer (standard or adaptive) never retries.

**Fix:** a `throttleRetryClient` decorator around the Data API client retrying throttle-signal errors (`ThrottlingException`, `Rate exceeded`, `SlowDown`) with exponential backoff + full jitter on all five Data API calls (paginator included). Attempt cap of 8 (~30s worst-case sleep), deliberately below the SDK's `retryMaxAttempts` since the layers can stack and the statement timeout is unset by default.

**DML-safety of retrying the submit calls** (`ExecuteStatement`/`BatchExecuteStatement`):
1. Only explicit throttle **responses** are retried — the server answered "rejected before acceptance". Ambiguous transport errors are never retried by this decorator.
2. Belt-and-braces: the **idempotency token is pinned before the retry loop**. The SDK auto-fills `ClientToken` per invocation (so its internal retries share one), but each decorator re-invocation would mint a fresh token — pinning it makes the service dedupe re-submissions even in ambiguous edge cases. Covered by a test asserting one token across all attempts.

Unit tests: recovery after wrapped throttles, exhaustion, non-throttle passthrough, context cancellation, token stability.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	for attempt := 0; attempt < c.maxAttempts; attempt++ {
+		err = call()
+		if !isWrappedThrottleError(err) || ctx.Err() != nil {
+			return err
+		}
+		// full jitter: sleep a random fraction of the current backoff ceiling
+		sleep := time.Duration(rand.Int63n(int64(delay)) + int64(delay)/2)
+		select {
+		case <-ctx.Done():
+			return err
```
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go]: Fixed in abd90eb43: cancellation during backoff (or detected after a throttled call) now returns an error wrapping `ctx.Err()` — `errors.Is(err, context.Canceled/DeadlineExceeded)` works — while keeping the last throttle text for diagnostics. Test asserts both properties.

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
```
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go]: Fixed in abd90eb43: cancellation during backoff (or detected after a throttled call) now returns an error wrapping `ctx.Err()` — `errors.Is(err, context.Canceled/DeadlineExceeded)` works — while keeping the last throttle text for diagnostics. Test asserts both properties.

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+// ExecuteStatement and BatchExecuteStatement SUBMIT statements, so retrying
+// them must never double-run DML. Two layers make that safe:
+//   - only explicit throttle RESPONSES are retried (the server answered
+//     "rejected before acceptance"); ambiguous transport errors are not
+//   - the idempotency token is pinned BEFORE the retry loop. The SDK auto-fills
+//     ClientToken per invocation, so its own internal retries share one token,
+//     but each decorator re-invocation would get a fresh one — pinning it here
+//     makes the service dedupe re-submissions even in ambiguous edge cases
+func (c *throttleRetryClient) ExecuteStatement(ctx context.Context, params *redshiftdata.ExecuteStatementInput, optFns ...func(*redshiftdata.Options)) (out *redshiftdata.ExecuteStatementOutput, err error) {
+	if params.ClientToken == nil {
```
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go:101]: Fixed in abd90eb43: nil inputs are initialized exactly like the generated SDK methods do before the ClientToken pinning, keeping the wrapper transparent. Covered by a nil-params test for both submit methods.

[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go:101]: Fixed in abd90eb43: nil inputs are initialized exactly like the generated SDK methods do before the ClientToken pinning, keeping the wrapper transparent. Covered by a nil-params test for both submit methods.

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go]: Already fixed in abd90eb43 (duplicate of the earlier resolved thread): cancellation returns an error wrapping `ctx.Err()` — `errors.Is`-detectable — with the last throttle text preserved.

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go:101]: Already fixed in abd90eb43 (duplicate of the earlier resolved thread): nil inputs are initialized like the generated SDK methods before the ClientToken pinning; nil-params test covers both submit methods.

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	for attempt := 0; attempt < c.maxAttempts; attempt++ {
+		err = call()
+		if !isWrappedThrottleError(err) {
+			return err
+		}
+		if ctx.Err() != nil {
+			return fmt.Errorf("%w (last throttle error: %v)", ctx.Err(), err)
+		}
+		// full jitter: sleep a random fraction of the current backoff ceiling
+		sleep := time.Duration(rand.Int63n(int64(delay)) + int64(delay)/2)
```
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go]: Good catch — fixed in 52653788a: equal jitter in `[delay/2, delay]`, so a sleep never exceeds the backoff ceiling and the worst-case budget note holds (sum of capped delays ≈ 26s for 8 attempts).

File: bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+		err = call()
+		if !isWrappedThrottleError(err) {
+			return err
+		}
+		if ctx.Err() != nil {
+			return fmt.Errorf("%w (last throttle error: %v)", ctx.Err(), err)
+		}
+		// equal jitter in [delay/2, delay]: randomized to decorrelate concurrent
+		// clients, but never exceeding the current backoff ceiling
+		sleep := delay/2 + time.Duration(rand.Int63n(int64(delay/2)+1))
```
[the remark this answers is not in the document]
Comment (absorbb) [reply bulker/bulkerlib/implementations/sql/redshift_driver/throttle.go:74]: Fixed: the loop now breaks before sleeping when the final attempt fails — terminal failures return immediately, and the wait budget (sum of 7 capped sleeps ≈ 26s max) matches the comment.
